File: zy70822/app/api/report_routes.py
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from typing import Dict, Any, Optional
from io import BytesIO

from app.models import get_db, RecordStatus
from app.services import ReportService

router = APIRouter()


@router.get("/summary", response_model=Dict[str, Any])
async def get_summary_report(
    batch_id: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """获取对账汇总报告"""
    try:
        service = ReportService(db)
        summary = service.generate_summary_report(batch_id)
        
        return {
            "success": True,
            "data": summary
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/details", response_model=Dict[str, Any])
async def get_detailed_records(
    batch_id: Optional[str] = None,
    status: Optional[RecordStatus] = None,
    db: Session = Depends(get_db)
):
    """获取对账明细记录"""
    try:
        service = ReportService(db)
        details = service.generate_detailed_records(
            batch_id=batch_id,
            status=status.value if status else None
        )
        
        return {
            "success": True,
            "total": len(details),
            "data": details
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/discrepancies", response_model=Dict[str, Any])
async def get_discrepancy_report(
    batch_id: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """获取差异分析报告"""
    try:
        service = ReportService(db)
        report = service.generate_discrepancy_report(batch_id)
        
        return {
            "success": True,
            "data": report
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/export/excel")
async def export_excel(
    batch_id: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """导出Excel对账报告"""
    try:
        service = ReportService(db)
        excel_data = service.export_to_excel(batch_id)
        
        output = BytesIO(excel_data)
        
        return StreamingResponse(
            output,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={
                "Content-Disposition": f"attachment; filename=vaccine_reconciliation_report.xlsx"
            }
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: zy70902/src/routes/reportRoutes.ts
import { Router, Request, Response } from 'express';
import { reportService } from '../services/reportService';

const router = Router();

router.post('/', async (req: Request, res: Response) => {
  try {
    const { batchId, resultId, generatedBy, periodStart, periodEnd } = req.body;

    if (!batchId || !resultId || !generatedBy) {
      return res.status(400).json({
        success: false,
        message: '缺少必要参数',
      });
    }

    const report = await reportService.generateReport(
      batchId,
      resultId,
      generatedBy,
      periodStart,
      periodEnd
    );

    res.json({
      success: true,
      data: report,
      message: '报告生成成功',
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: '报告生成失败',
      error: error instanceof Error ? error.message : String(error),
    });
  }
});

router.get('/:reportId', async (req: Request, res: Response) => {
  try {
    const { reportId } = req.params;
    const report = reportService.getReport(reportId);

    if (!report) {
      return res.status(404).json({
        success: false,
        message: '报告不存在',
      });
    }

    res.json({
      success: true,
      data: report,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: '获取报告失败',
      error: error instanceof Error ? error.message : String(error),
    });
  }
});

router.get('/', async (req: Request, res: Response) => {
  try {
    const reports = reportService.getAllReports();

    res.json({
      success: true,
      data: reports,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: '获取报告列表失败',
      error: error instanceof Error ? error.message : String(error),
    });
  }
});

router.get('/:reportId/excel', async (req: Request, res: Response) => {
  try {
    const { reportId } = req.params;
    const buffer = await reportService.generateExcelReport(reportId);

    res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
    res.setHeader('Content-Disposition', `attachment; filename="reconciliation-report-${reportId}.xlsx"`);
    res.send(buffer);
  } catch (error) {
    res.status(500).json({
      success: false,
      message: '生成Excel报告失败',
      error: error instanceof Error ? error.message : String(error),
    });
  }
});

router.get('/:reportId/pdf', async (req: Request, res: Response) => {
  try {
    const { reportId } = req.params;
    const buffer = await reportService.generatePdfReport(reportId);

    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="reconciliation-report-${reportId}.pdf"`);
    res.send(buffer);
  } catch (error) {
    res.status(500).json({
      success: false,
      message: '生成PDF报告失败',
      error: error instanceof Error ? error.message : String(error),
    });
  }
});

export default router;
